aether-kernel/aether/kernel/api/utils.py

```diff
@@ -52,3 +56,125 @@ def merge_objects(source, target, direction):
     else:
         result = target
     return result
+
+
+def get_unique_schemas_used(mappings_ids):
+    result = {}
+    schemas = models.Schema.objects.filter(schemadecorators__mappings__id__in=mappings_ids)
+    for schema in schemas:
+        other_linked_mappings = models.Mapping.objects.filter(
+            schemadecorators__schema__id=schema.id
+        ).exclude(id__in=mappings_ids)
+        result[schema.name] = {
+            'id': schema.id,
+            'name': schema.definition['name'],
+        }
+        if other_linked_mappings:
+            result[schema.name]['is_unique'] = False
+        else:
+            result[schema.name]['is_unique'] = True
+    return result
+
+
+@transaction.atomic
+def bulk_delete_by_mappings(delete_opts={}, mappingset=None, mappings=[]):
+    '''
+    Bulk delete submissions, entities and schemas uniquely
+    linked to the supplied mappings or mappingset
+
+    Arguments:
+
+    delete_opts: {
+        # dict of artefacts to be deleted
+
+        'entities': True | False,
+        'schemas': True | False,
+        'submissions': True | False
+    },
+
+    # the uuid of the mappingset to delete it's artefacts.
+    # No need to provide mappings if mappingset exists.
+    mappingset: 'uuid',
+
+    mappings:  [
+        # a list of mapping uuids to delete related artefacts,
+
+        'uuid-1',
+        'uuid-2'
+    ]
+
+    returned result: {
+        # a dict of deleted artefacts
+
+        'entities': 34 (int)
+        # Number of deleted entities
+
+        'submissions': 34 (int)
+        # Number of deleted submissions
+
+        'schemas': {
+
+            # schemas used in the context of the supplied mappings
+            # with details on delete operations perform on them or not.
+            'Schema-1': {
+                # id of the schema
+                'id': uuid,
+
+                # name of the schema definition (this can be different from the schema.name)
+                'name': string,
+
+                # Flag indicating if the schema is uniquely used in the mappings' context.
+                # Delete operations will not be performed if False
+                'is_unique': True | False,
+
+                # Flag if delete operation was successful on this schema and cascading objects.
+                # Objects with {'is_unique':False} will not have this property
+                'is_deleted': True | False,
+
+                # Reason for a delete failure, only available if {'is_deleted': False}
+                'reason': string
+            },
+            'Schema-2': {
+                ...
+            },
+            ...
+        }
+
+    }
+    '''
+
+    if mappingset:
+        mappings = models.Mapping.objects.filter(mappingset=mappingset).values_list('id', flat=True)
```

Comment:
Here you are overwriting the given `mappings` list, does this mean that or the mappingset or the mappings list but not both?

Reply:
also... the shortcut `mappingset.mappings.values_list('id', flat=True)` ;-)

then 
```python
mappings += mappingset.mappings.values_list('id', flat=True)
```

Reply:
It means if a mappingset is provided, use only the mappings within that mappingset. Thus, overwrite the provided mappings if any

Reply:
also about the shortcut, `mappingset` passed is a string not ofType `MappingSet`